### ISSUE TITLE
[CUDA] fake_quantize_per_channel_affine gives incorrect result for extreme finite float32 input

### ISSUE THREAD

**Issue by tangyuan499-commits** [pytorch/pytorch#195343]
### 🐛 Describe the bug

torch.fake_quantize_per_channel_affine produces different results on CPU and CUDA for an extreme but finite float32 input with a small per-channel scale.

On PyTorch 2.13.0+cu126, the CPU result for the first element is exactly 0.0, while CUDA returns approximately 2.55e-05.

Minimal reproducer:

```
import torch

x = torch.tensor(
    [[-3.4028235e38, -1.0e-6],
     [1.0e6, 3.4028235e38]],
    dtype=torch.float32,
)
scale = torch.tensor([1.0e-7, 1.0e20], dtype=torch.float32)
zero_point = torch.tensor([-128, 127], dtype=torch.int32)

cpu = torch.fake_quantize_per_channel_affine(
    x, scale, zero_point, 0, -128, 127
)

cuda = torch.fake_quantize_per_channel_affine(
    x.cuda(), scale.cuda(), zero_point.cuda(), 0, -128, 127
).cpu()

print("CPU:")
print(cpu)
print("CUDA:")
print(cuda)
print("Exact equal:", torch.equal(cpu, cuda))

print("CPU bits:",
      [f"0x{v & 0xffffffff:08x}"
       for v in cpu.contiguous().view(torch.int32).flatten().tolist()])
print("CUDA bits:",
      [f"0x{v & 0xffffffff:08x}"
       for v in cuda.contiguous().view(torch.int32).flatten().tolist()])
```

Observed output:

CPU:
tensor([[0., 0.],
        [0., 0.]])

CUDA:
tensor([[2.5500e-05, 0.0000e+00],
        [0.0000e+00, 0.0000e+00]])

Exact equal: False

CPU bits:
['0x00000000', '0x00000000', '0x00000000', '0x00000000']

CUDA bits:
['0x37d5e8d5', '0x00000000', '0x00000000', '0x00000000']

With higher print precision, the first CUDA value is:

2.54999995377147570252e-05
Why this does not appear to be an expected numerical-accuracy difference

For the first element,

input      = -3.4028235e38
scale      = 1e-7
zero_point = -128
quant_min  = -128
quant_max  = 127

the fake-quantized value is far below the representable quantization range and should therefore saturate to quant_min = -128.

The dequantized result should consequently be:

(-128 - (-128)) * 1e-7 = 0.0

which is the CPU result.

The CUDA result is instead approximately:

2.55e-05 = (127 - (-128)) * 1e-7

so the discrepancy spans the full quantization range rather than being a small floating-point rounding difference.

The raw float32 representations also differ (0x00000000 vs 0x37d5e8d5), so this is not caused by tensor printing precision.

This may be related to #171777, which addressed a similar issue for CUDA fake quantization in fake_quantize_per_tensor_affine. This reproducer affects fake_quantize_per_channel_affine and uses finite input values.

### Versions

Collecting environment information...
PyTorch version: 2.13.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 4.3.4
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-4.15.0-213-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Versions of relevant libraries:
[pip3] numpy==2.5.2
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] optree==0.19.1
[pip3] torch==2.13.0+cu126
[pip3] torchaudio==2.11.0+cu126
[pip3] torchelastic==0.2.2
[pip3] torchvision==0.28.0+cu126
[pip3] triton==3.7.1

cc @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

**Comment by priyanshusky0** [comment]
@tangyuan499-commits , Are you working on this issue? if not i am happy to work on this !
